for-research/public-likelihoods/public-likelihoods.ipynb

```diff
@@ -0,0 +1,936 @@
+   ],
+   "source": [
+    "w[\"datasets/obsData\"].SetChecked();w[\"datasets/asimovData\"].SetChecked(False) # selects which datasets to display with the pdf\n",
+    "ROOT.gStyle.SetOptStat(False) # turn off the stats-box that ROOT will display by default on every plot\n",
+    "c = ROOT.TCanvas(\"canv\",\"canv\",900,900)\n",
+    "ROOT.gStyle.SetLegendFont(43);ROOT.gStyle.SetLegendTextSize(12);\n",
+    "w[\"models/simPdf\"].Draw(\"eratio\") # draw model with error bars and ratio plot\n",
+    "c.Draw()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cb976100-622b-4141-b822-d0fe17342559",
+   "metadata": {},
+   "source": [
+    "# 2. Modifying workspaces\n",
+    "\n",
+    "Workspaces can be modified 'in-situ' or a new workspace can be created and objects added to it. We will see an example of both of these now ...\n",
+    "\n",
+    "## 2.1. Modifying in-situ: Reparameterizing a parameter\n",
+    "\n",
+    "We can use the `xRooNode::Replace` method to replace any variable with a generic function. The function must follow RooFit's [WS-factory language](https://root.cern.ch/doc/master/classRooFactoryWSTool.html#a228d9cb336a3bdc7dfe260f4c6aaf8df). For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "1f9695ee-06d0-4fdd-88da-e6f0c4ea98e1",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "xRooNode{}"
+      ]
+     },
+     "execution_count": 6,
+     "metadata": {},
+     "output_type": "execute_result"
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Warning in <xRooNode::Replace>: Replacing dPhi_ttbar_Z_Bin_001_mu in all clients: dPhi_ttbar_Z_Bin_005_mu,mu_prime_2,model_reg_3l_WZ_dPhi_ttbar_Z_reg_3l_WZ_Truth_bin_1_scaleFactors,model_reg_3l_tZq_dPhi_ttbar_Z_reg_3l_tZq_Truth_bin_1_scaleFactors,model_reg_4l_DF_dPhi_ttbar_Z_reg_4l_DF_Truth_bin_1_scaleFactors,model_reg_3l_ttZ_dPhi_ttbar_Z_reg_3l_ttZ_Truth_bin_1_scaleFactors,model_reg_4l_SF_dPhi_ttbar_Z_reg_4l_SF_Truth_bin_1_scaleFactors,model_reg_4l_ZZ_dPhi_ttbar_Z_reg_4l_ZZ_Truth_bin_1_scaleFactors,\n",
+      "Info in <xRooNode::SaveAs>: workspace saved to reparameterized_ws.root\n",
+      "Warning in <xRooNode::Replace>: Replacing myFunc in all clients: dPhi_ttbar_Z_Bin_005_mu,mu_prime_2,model_reg_3l_WZ_dPhi_ttbar_Z_reg_3l_WZ_Truth_bin_1_scaleFactors,model_reg_3l_tZq_dPhi_ttbar_Z_reg_3l_tZq_Truth_bin_1_scaleFactors,model_reg_4l_DF_dPhi_ttbar_Z_reg_4l_DF_Truth_bin_1_scaleFactors,model_reg_3l_ttZ_dPhi_ttbar_Z_reg_3l_ttZ_Truth_bin_1_scaleFactors,model_reg_4l_SF_dPhi_ttbar_Z_reg_4l_SF_Truth_bin_1_scaleFactors,model_reg_4l_ZZ_dPhi_ttbar_Z_reg_4l_ZZ_Truth_bin_1_scaleFactors,\n"
+     ]
+    }
+   ],
+   "source": [
+    "poiName = w.poi()[0].GetName() # get the name of the first POI\n",
+    "w.poi()[0].Replace(f\"expr::myFunc('2*{poiName}',{poiName})\") # example of replacing the parameter everywhere in the workspace with a function that multiplies it by 2\n",
```

Comment:
I'm not sure even I can follow the formatting in this code. `Replace()` takes a string, and `myFunc()` takes a string that you're using for the replacement and a name (not a string) that you're replacing? (poiName, the second parameter to myFunc, is not in single quotes here) If that's right... wow we made something tricky :(

Reply:
I think all of the bit you find confusing is relating to the RooFit factory syntax, which yes is unfortunately a domain specific language. Do you have a suggestion of how to make it clearer? That line is creating a RooFormularVar called myFunc, which takes on twice the value of the parameter of interest, and then replacing the parameter of interest in the workspace with that function, so that it replaces everywhere the poi appeared in any pdfs.

Reply:
I'm trying to track this through here, but I must be getting hung up in your classes. `w` is an `xRooNode` (from L79 above), right? So I should look here?
https://gitlab.cern.ch/will/xroofit/-/blob/master/xRooFit/xRooNode.h?ref_type=heads

Then `w.poi()` returns an `xRooNode`, and `operator[]` also returns an `xRooNode`? (It feels like turtles all the way down with this thing)  `Replace` then is your function that takes another xRooNode:
https://gitlab.cern.ch/will/xroofit/-/blob/master/xRooFit/xRooNode.h?ref_type=heads#L315

which hits one of the constructors with a bunch of default parameters... which? I don't want to trace this through now. I have to say, unless this is super well documented, as soon as you start trying to see what the class is doing this is a very, very confusing setup...

I hear you on RooFit factory syntax being ugly, but I think we have to strive to teach people how to do things and why, and try to avoid teaching them magic spells that they repeat without any understanding of the syntax.